corehq/apps/userreports/views.py

```diff
@@ -807,40 +880,103 @@ def create_data_source_from_app(request, domain):
                 data_source.save()
                 messages.success(request, _(u"Data source created for '{}'".format(xform.default_name())))
 
-            return HttpResponseRedirect(reverse('edit_configurable_data_source', args=[domain, data_source._id]))
-    else:
-        form = ConfigurableDataSourceFromAppForm(domain)
-    context = _shared_context(domain)
-    context['sources_map'] = form.app_source_helper.all_sources
-    context['form'] = form
-    return render(request, 'userreports/data_source_from_app.html', context)
-
-
-def _edit_data_source_shared(request, domain, config, read_only=False, create=False):
-    if request.method == 'POST':
-        form = ConfigurableDataSourceEditForm(domain, config, read_only, data=request.POST)
-        if form.is_valid():
-            config = form.save(commit=True)
-            messages.success(request, _(u'Data source "{}" saved!').format(config.display_name))
-            if create:
-                # if we just created a data source, redirect to the edit view to avoid creating a new one
-                return HttpResponseRedirect(reverse('edit_configurable_data_source', args=[domain, config._id]))
-    else:
-        form = ConfigurableDataSourceEditForm(domain, config, read_only)
-    context = _shared_context(domain)
-    context.update({
-        'form': form,
-        'data_source': config,
-        'read_only': read_only
-    })
-    if config.is_deactivated:
-        messages.info(
-            request, _(
-                'Data source "{}" has no associated table.\n'
-                'Click "Rebuild Data Source" to recreate the table.'
-            ).format(config.display_name)
+            return HttpResponseRedirect(reverse(
+                EditDataSourceView.urlname, args=[self.domain, data_source._id]
+            ))
+        return self.get(request, *args, **kwargs)
+
+    @property
+    def page_context(self):
+        return {
+            'sources_map': self.form.app_source_helper.all_sources,
+            'form': self.form,
+        }
+
+
+class BaseEditDataSourceView(BaseUserConfigReportsView):
+    template_name = 'userreports/edit_data_source.html'
+
+    @property
+    def page_context(self):
+        return {
+            'form': self.edit_form,
+            'data_source': self.config,
+            'read_only': self.read_only
+        }
+
+    @property
+    def page_url(self):
+        if self.config_id:
+            return reverse(self.urlname, args=(self.domain, self.config_id,))
+        return super(BaseEditDataSourceView, self).page_url
+
+    @property
+    def config_id(self):
+        return self.kwargs.get('config_id')
+
+    @property
+    def read_only(self):
+        return id_is_static(self.config_id) if self.config_id is not None else False
+
+    @property
+    @memoized
+    def config(self):
+        if self.config_id is None:
+            return DataSourceConfiguration(domain=self.domain)
+        try:
```

Comment:
same question as above - any reason this can't use `get_datasource_config_or_404`?